### PR TITLE
Add QosApi

### PR DESCRIPTION
**Pull request by yzhao174**
Add SDK support for pinging each region and get RTT from the regions

### CODE REVIEW

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+            QosResult result = await GetResultAsync(timeoutMs);
+            if (result.ErrorCode != 0)
+            {
+                return result;
+            }
+
+            bool eventSent = await SendResultsToPlayFab(result);
+            if (!eventSent)
+            {
+                result.ErrorCode = -3;
```
Comment (toddbellMSFT) [inline PlayFabSDK/source/Qos/PlayFabQosApi.cs]: What are the consequences of the error code being -3? What does that indicate?

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+                return result;
+            }
+
+            // get datacenter map (call thunderhead)
+            await GetQoSServerList();
+
+            int serverCount = _dataCenterMap.Count;
+            if (serverCount < 0)
+            {
+                result.ErrorCode = -2;
```
Comment (toddbellMSFT) [inline PlayFabSDK/source/Qos/PlayFabQosApi.cs]: Can we set these numbers to an enum (then cast to an int for the error code) or named const int so that we can read specifically what went wrong here?

Comment (toddbellMSFT) [reply PlayFabSDK/source/Qos/PlayFabQosApi.cs]: We can at least also LogQos, but having a stack that is understandable is also a nice thing to aim for if the logger is optional

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                // If the dataCenterMap is already initialized, return
+                return;
+            }
+
+            var request = new ListQosServersRequest();
+            PlayFabResult<ListQosServersResponse> response = await PlayFabMultiplayerAPI.ListQosServersAsync(request);
+            
+            if (response?.Error != null)
+            {
+                LogQos("Could not get the server list from thunderhead\n Error : " + response.Error.ErrorMessage);
```
Comment (toddbellMSFT) [inline PlayFabSDK/source/Qos/PlayFabQosApi.cs]: Should this method exit if there's an error now? Or can we guarantee response.Result.QosServers will get called correctly?

File: PlayFabSDK/source/Qos/RegionPinger.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            List<Task<int>> latencyMeasures = Enumerable.Range(0, PingCount).Select(i => PingInternalAsync()).ToList();
+            await Task.WhenAny(Task.WhenAll(latencyMeasures), timeout);
+            List<int> latencies = latencyMeasures.Where(t => t.IsCompleted && t.Result >= 0).Select(t => t.Result).ToList();
+
+            // Return the average of the remaining numbers
+            return new QosRegionResult
+            {
+                Region = _region,
+                LatencyMs = latencies.Any()? (int) latencies.Average() : UnknownLatencyValue,
+                ErrorCode = latencyMeasures.All(t => t.IsCompleted) ? 0 : -1
```
Comment (toddbellMSFT) [inline PlayFabSDK/source/Qos/RegionPinger.cs]: Can we get a const for the ErrorCode here? Or a LogQoS?

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs (hunk 1 above)
Comment (pgilmorepf) [reply PlayFabSDK/source/Qos/PlayFabQosApi.cs]: /Agree, you can do something simple like
```const int MEANINGFUL_REASON = -3```
but hard coded numbers are not awesome.

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs (hunk 2 above)
Comment (pgilmorepf) [reply PlayFabSDK/source/Qos/PlayFabQosApi.cs]: Enums are good too, I like enums.

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs (hunk 1 above)
Comment (yzhao174) [reply PlayFabSDK/source/Qos/PlayFabQosApi.cs]: Make sense. I will assign meaingful name to the integers.

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+                LogQos("Client is not logged in");
+                result.ErrorCode = -1;
+                return result;
+            }
+
+            // get datacenter map (call thunderhead)
+            await GetQoSServerList();
+
+            int serverCount = _dataCenterMap.Count;
+            if (serverCount < 0)
```
Comment (pgilmorepf) [inline PlayFabSDK/source/Qos/PlayFabQosApi.cs]: I feel like == 0 would be an error too.

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs (hunk 3 above)
Comment (pgilmorepf) [reply PlayFabSDK/source/Qos/PlayFabQosApi.cs]: response.Error.GenerateErrorReport() will provide a much more complete view of the data contained in the Error object (there is often more information than just ErrorMessage)

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs (hunk 2 above)
Comment (yzhao174) [reply PlayFabSDK/source/Qos/PlayFabQosApi.cs]: Will make the change

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs (hunk 3 above)
Comment (yzhao174) [reply PlayFabSDK/source/Qos/PlayFabQosApi.cs]: Yes, it is safe to just return from here.

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+            }
+
+            foreach (QosServer qosServer in response.Result.QosServers)
+            {
+                if (!qosServer.Region.HasValue)
+                {
+                    continue;
+                }
+
+                _dataCenterMap[qosServer.Region.Value] = qosServer.ServerUrl;
```
Comment (pgilmorepf) [inline PlayFabSDK/source/Qos/PlayFabQosApi.cs:85]: I feel like:
```
if (qosServer.Region.HasValue)
    _dataCenterMap[qosServer.Region.Value] = qosServer.ServerUrl;
```
Is a more understandable code-flow. The continue as an early out, but the "work" only being 1 line is a little weird.

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        private async Task<List<QosRegionResult>> GetSortedRegionLatencies(int timeoutMs)
+        {
+            List<Task<QosRegionResult>> asyncPingResults = _dataCenterMap.Select(async kvp =>
+            {
+                var regionPinger = new RegionPinger(kvp.Value, kvp.Key);
+                return await regionPinger.PingAsync(timeoutMs);
+            }).ToList();
+
+            await Task.WhenAll(asyncPingResults);
+            return asyncPingResults.Select(t => t.Result).OrderBy(t => t.LatencyMs).ToList();
```
Comment (pgilmorepf) [inline PlayFabSDK/source/Qos/PlayFabQosApi.cs]: Linq statements can produce some very cool and (if you understand Linq) easy to understand code.
Conversely, if you aren't familiar with Linq, this can be very hard to parse.
Linq is fast, but it also hits the garbage collector VERY hard, and games can be very sensitive to that.
For example, 5 Linq statements like this will actually allocate/churn MUCH more than 5 times the memory as the single List<QosRegionResult> you return.

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+        }
+
+        private async Task<bool> SendResultsToPlayFab(QosResult result)
```
Comment (pgilmorepf) [inline PlayFabSDK/source/Qos/PlayFabQosApi.cs]: Design wise, is the bool return on this meaningful?
It seems like it could just be a fire-and-forget that executes a task async, without blocking the caller.

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+            }
+            catch (Exception exception)
+            {
+                LogQos(exception.ToString());
+                return false;
+            }
+
+            if (response == null || response.Error != null)
+            {
+                LogQos(response?.Error.ErrorMessage);
```
Comment (pgilmorepf) [inline PlayFabSDK/source/Qos/PlayFabQosApi.cs]: response?.Error.GenerateErrorReport()

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+        private static async Task<PlayFabResult<WriteEventsResponse>> WriteTelemetryAsync(WriteEventsRequest request, object customData = null, Dictionary<string, string> extraHeaders = null)
+        {
+            if ((request?.AuthenticationContext?.EntityToken ?? PlayFabSettings.staticPlayer.EntityToken) == null) throw new PlayFabException(PlayFabExceptionCode.EntityTokenNotSet, "Must call GetEntityToken before calling this method");
+
+            var httpResult = await PlayFabHttp.DoPost("/Event/WriteTelemetryEvents", request, "X-EntityToken", PlayFabSettings.staticPlayer.EntityToken, extraHeaders);
+            if (httpResult is PlayFabError)
+            {
+                var error = (PlayFabError)httpResult;
+                PlayFabSettings.GlobalErrorHandler?.Invoke(error);
```
Comment (pgilmorepf) [inline PlayFabSDK/source/Qos/PlayFabQosApi.cs:147]: GlobalErrorHandler existed when the only SDK feature was the API.
And, I'm not certain I want to continue the pattern as-is.
I think the QOS system could have its own ErrorHandler callback, but I think we should rename/phase-out the GlobalErrorHandler in general.  (The error type itself is very specific to the PlayFab API, and it's only a coincidence that it matches your error here).

Comment (pgilmorepf) [reply PlayFabSDK/source/Qos/PlayFabQosApi.cs:147]: GlobalErrorHandler is also not compatible with the general move towards instanced APIs and such.

File: PlayFabSDK/source/Qos/QosRegionResult.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
+﻿namespace PlayFab.QoS
+{
+    using MultiplayerModels;
+
+    public class QosRegionResult
+    {
+        public AzureRegion Region { get; set; }
```
Comment (pgilmorepf) [inline PlayFabSDK/source/Qos/QosRegionResult.cs:7]: A { get; set; } property is syntactically identical to a field, but about 1000x more expensive than fields.
I suggest only using properties when there is a meaningful reason to do so (inheritance for example).

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs (hunk 6 above)
Comment (yzhao174) [reply PlayFabSDK/source/Qos/PlayFabQosApi.cs:85]: I try to early out everywhere in this function and classes. I think that's more consistent. However, if that is a project coding style, I can change it.

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs (hunk 10 above)
Comment (yzhao174) [reply PlayFabSDK/source/Qos/PlayFabQosApi.cs:147]: Here I just create a simple call to WriteTelemetryEvents, since its API doesn't exist in SDK, and there is no plan to create that API in SDK. It is the same as other direct generated APIs. So I would want to keep it consistent with WriteEvents API call.

File: PlayFabSDK/source/Qos/QosRegionResult.cs (hunk 11 above)
Comment (yzhao174) [reply PlayFabSDK/source/Qos/QosRegionResult.cs:7]: Would you point me some reference about performance benchmark using property vs field? From my knowledge, an auto property is the same as a field in performance. Properties are usually exposed as public and field is more used privately for packaging and extension purpose.

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs (hunk 7 above)
Comment (yzhao174) [reply PlayFabSDK/source/Qos/PlayFabQosApi.cs]: I changed it to for-loop + sort.

File: PlayFabSDK/source/Qos/PlayFabQosApi.cs (hunk 8 above)
Comment (yzhao174) [reply PlayFabSDK/source/Qos/PlayFabQosApi.cs]: I think it makes sense to not block the client on emitting metrics. Updated
